chain/chain/src/chain.rs

```diff
@@ -1460,47 +1464,82 @@ impl Chain {
+        //});
+    }
+
+    /// Check if there are any block that finishes applying chunks. Run postprocessing on the block.
+    fn try_postprocess_block(
+        &mut self,
+        me: &Option<AccountId>,
+        block_accepted: &mut dyn FnMut(AcceptedBlock),
+    ) -> Result<Option<Tip>, Error> {
+        if let Ok((block_hash, apply_results)) = self.apply_chunks_receiver.try_recv() {
+            let (block, block_preprocess_info) =
+                self.processing_blocks.take_preprocess_block(&block_hash).unwrap();
+            let prev_head = self.store.head()?;
+            let mut chain_update = self.chain_update();
+            let provenance = block_preprocess_info.provenance.clone();
+            let new_head =
+                chain_update.postprocess_block(me, &block, block_preprocess_info, apply_results)?;
+            chain_update.chain_store_update.save_block_height_processed(block.header().height());
+            chain_update.commit()?;
 
-        if let Some(tip) = &new_head {
-            if let Ok(producers) = self
-                .runtime_adapter
-                .get_epoch_block_producers_ordered(&tip.epoch_id, &tip.last_block_hash)
-            {
-                let mut count = 0;
-                let mut stake = 0;
-                for (info, is_slashed) in producers.iter() {
-                    if !*is_slashed {
-                        stake += info.stake();
-                        count += 1;
+            self.pending_states_to_patch = None;
+
+            if let Some(tip) = &new_head {
+                if let Ok(producers) = self
+                    .runtime_adapter
+                    .get_epoch_block_producers_ordered(&tip.epoch_id, &tip.last_block_hash)
+                {
+                    let mut count = 0;
+                    let mut stake = 0;
+                    for (info, is_slashed) in producers.iter() {
+                        if !*is_slashed {
+                            stake += info.stake();
+                            count += 1;
+                        }
                     }
+                    stake /= NEAR_BASE;
+                    metrics::VALIDATOR_AMOUNT_STAKED.set(i64::try_from(stake).unwrap_or(i64::MAX));
+                    metrics::VALIDATOR_ACTIVE_TOTAL.set(count);
                 }
-                stake /= NEAR_BASE;
-                metrics::VALIDATOR_AMOUNT_STAKED.set(i64::try_from(stake).unwrap_or(i64::MAX));
-                metrics::VALIDATOR_ACTIVE_TOTAL.set(count);
-            }
-        };
+            };
 
-        let status = self.determine_status(new_head.clone(), prev_head);
+            let status = self.determine_status(new_head.clone(), prev_head);
 
-        // Notify other parts of the system of the update.
-        block_accepted(AcceptedBlock { hash: *block.hash(), status, provenance });
+            // Notify other parts of the system of the update.
+            block_accepted(AcceptedBlock { hash: *block.hash(), status, provenance });
```

Comment:
Oh.... This callback might prove to be challenging to handle, as we'll completely change the place where it is called once this is async. 

Not sure what's the best local fix here, but in general am not a fan of such "push" callbacks:

* the callback has locked-down effects -- it can't return an error, can't be async
* calling user-supplied code with our stuff on the stack is icky (invariant might be temporarily broken)
* this often doesn't play well with lifetimes (because lifetimes protect you from observing temporarily broken invariants)

I prefer the pattern where the caller polls for what happend since last time.

```rust
// Pushing events via observer -- no
pub struct StatefulThing {
   ...
}

pub struct Event;

impl StatefulThing {
  pub process(&mut self, callback: &mut dyn FnMut(Event));
}

// Pulling events via changes buffer -- Yes!
pub struct StatefulThing {
   new_events: Vec<Event>,
}

pub struct Event;

impl StatefulThing {
  pub fn process(&mut self);
  pub fn take_events(&mut self) -> Vec<Event> { std::mem::take(&mut self.new_events) }
}
```

Use this a bunch in rust-analyzer, eg here:

https://github.com/rust-lang/rust-analyzer/blob/90236dd77ae817b5e2c6d30b3523890975d4d519/crates/rust-analyzer/src/main_loop.rs#L474-L476